src/Compilers/CSharp/Test/Semantic/Diagnostics/GetDiagnosticsTests.cs

```diff
@@ -694,5 +695,43 @@ public async Task TestGetAnalyzerSyntaxDiagnosticsWithCancellation(bool concurre
             var diagnostic = Assert.Single(diagnostics);
             Assert.Equal(RegisterSyntaxTreeCancellationAnalyzer.DiagnosticId, diagnostic.Id);
         }
+
+        [Fact]
+        public async Task TestEventQueuePartialCompletionForSpanBasedQuery()
+        {
+            var source = @"
+class C
+{
+    void M1()
+    {
+        int x1 = 0;
+    }
+
+    void M2()
+    {
+        int x2 = 0;
+    }
+}";
+            var compilation = CreateCompilation(source);
+            var syntaxTree = compilation.SyntaxTrees[0];
+            var semanticModel = compilation.GetSemanticModel(syntaxTree);
+
+            // Get analyzer diagnostics for a span within "M1".
+            var localDecl = syntaxTree.GetRoot().DescendantNodes().OfType<LocalDeclarationStatementSyntax>().First();
+            var span = localDecl.Span;
+            var analyzers = ImmutableArray.Create<DiagnosticAnalyzer>(new CSharpCompilerDiagnosticAnalyzer());
+            var compilationWithAnalyzers = compilation.WithAnalyzers(analyzers, AnalyzerOptions.Empty);
+            _ = await compilationWithAnalyzers.GetAnalysisResultAsync(semanticModel, span, CancellationToken.None);
+
+            // Verify only required compilation events are generated in the event queue.
+            // Event queue should not be completed as we are requesting diagnostics for a span within "M1"
+            // and no compilation event should be generated for "M2".
+            var eventQueue = compilationWithAnalyzers.Compilation.EventQueue;
+            Assert.False(eventQueue.IsCompleted);
+
+            // Now fetch diagnostics for entire tree and verify event queue is completed.
+            _ = await compilationWithAnalyzers.GetAnalysisResultAsync(semanticModel, filterSpan: null, CancellationToken.None);
+            Assert.True(eventQueue.IsCompleted);
+        }
```

Comment:
can we have a test that validates that if we ask for a span inside a method, that SemanticSpan analyzers only run for that method?

Reply:
Let me create a separate IDE PR for all the documentation that is being requested as well as this additional test. I'd like to keep this PR small for the core compiler layer change and test for ease of review for the compiler team.

Reply:
i woudl at least doc the code change with just a few lines explaining the point of passing that span along :)

Reply:
(but i'm otherwise ok with full docs coming in a separate PR).

Reply:
Done - added a comment to this PR and also opened a separate IDE PR https://github.com/dotnet/roslyn/pull/56751 for the suggested IDE test and documentation